lib/seam/paginator.rb

```diff
@@ -0,0 +1,106 @@
+# frozen_string_literal: true
+
+require "faraday"
+require_relative "http"
+
+module Seam
+  THREAD_CONTEXT_KEY = :seam_pagination_context
+  PaginationContext = Struct.new(:pagination_data)
+
+  class Paginator
+    def initialize(request, params = {})
+      raise ArgumentError, "request must be a Method" unless request.is_a?(Method)
+      raise ArgumentError, "params must be a Hash" unless params.is_a?(Hash)
+
+      @request = request
+      @params = params.transform_keys(&:to_sym)
+    end
+
+    def first_page
+      fetch_page(@params)
+    end
+
+    def next_page(next_page_cursor)
+      if next_page_cursor.nil? || next_page_cursor.empty?
+        raise ArgumentError,
+          "Cannot get the next page with a nil or empty next_page_cursor."
+      end
+
+      fetch_page(@params.merge(page_cursor: next_page_cursor))
+    end
+
+    def flatten_to_list
+      all_items = []
+      current_items, pagination = first_page
+
+      all_items.concat(current_items) if current_items
+
+      while pagination&.has_next_page? && (cursor = pagination.next_page_cursor)
+        current_items, pagination = next_page(cursor)
+        all_items.concat(current_items) if current_items
+      end
+
+      all_items
+    end
+
+    def flatten
+      Enumerator.new do |yielder|
+        current_items, pagination = first_page
+        current_items&.each { |item| yielder << item }
+
+        while pagination&.has_next_page? && (cursor = pagination.next_page_cursor)
+          current_items, pagination = next_page(cursor)
+          current_items&.each { |item| yielder << item }
+        end
+      end
+    end
+
+    private
+
+    def fetch_page(params)
+      context = PaginationContext.new(nil)
+      Thread.current[THREAD_CONTEXT_KEY] = context
```

Comment:
I went with this approach because we can't implement something similar to Python's implementation. Faraday's middleware builder locks its stack after it's been initialized, and you cannot insert or remove middleware on a live connection.

Reply:
This is honestly a better approach anyway! Adding are removing the middlware dynamically (my original suggestion) looks fragile in hindsight. Better to always attach the middleware and use a cache / context like this.